### PR TITLE
8289196: Pattern domination not working properly for record patterns

### PR DESCRIPTION
**Pull request by lahodaj**
When checking that a pattern in a switch case is not dominated by a preceding pattern, handling for record pattern is mostly missing. This patch is attempting to fix that, and as it requires a bit more code, it moves the check into `Check`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289196](https://bugs.openjdk.org/browse/JDK-8289196): Pattern domination not working properly for record patterns

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [220fcc12](https://git.openjdk.org/jdk19/pull/84/files/220fcc12606d35abfa4467dd877bbfa5987b3872)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk19 pull/84/head:pull/84` \
`$ git checkout pull/84`

Update a local copy of the PR: \
`$ git checkout pull/84` \
`$ git pull https://git.openjdk.org/jdk19 pull/84/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 84`

View PR using the GUI difftool: \
`$ git pr show -t 84`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk19/pull/84.diff">https://git.openjdk.org/jdk19/pull/84.diff</a>

</details>

### CODE REVIEW

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java [hunk 1]
```diff
@@ -4373,6 +4373,95 @@ void checkSwitchCaseStructure(List<JCCase> cases) {
...
+            });
+        }
+        private boolean patternDominated(JCPattern existingPattern, JCPattern currentPattern) {
+            Type existingPatternType = types.erasure(existingPattern.type);
+            Type currentPatternType = types.erasure(currentPattern.type);
+            if (existingPatternType.isPrimitive() ^ currentPatternType.isPrimitive()) {
+                return false;
+            }
+            if (existingPatternType.isPrimitive()) {
+                if (!types.isSameType(existingPatternType, currentPatternType)) {
```
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:4423]: note: shouldn't we return true if both are primitive and they are the same type? I mean this method seems to returns the right thing in that case but it is doing some extra analysis that doesn't seem to be necessary IMO

**Comment by vicente-romero-oracle** [comment]
shouldn't this test case be rejected too:

```
class TestRecordPattern {
    int testRecordPatternsDominated2() {
        record R(int a) {}
        Object o = null;
        switch (o) {
            //case R(int a): return 1;
            case R(int a) when guard(): return -1;
            case R(int a): return 1;
            default: return 0;
        }
    }

    boolean guard() {
        return false;
    }
}
```

http://cr.openjdk.java.net/~gbierman/jep427+405/jep427+405-20220601/specs/patterns-switch-record-patterns-jls.html#jls-14.30.3 doesn't mention anything about guards if I understand correctly

File: src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java (hunk 1 above)
Comment (lahodaj) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Check.java:4423]: We could, but I am not sure if that's not "too clever". Returning `true` when the type is same is based on the fact that only binding patterns (type test patterns) can be of primitive types, so the tests below will always return `true`. I can do that if strongly preferred, but I somewhat like the separation of concerns, where we verify the types here, and then the structure is checked below, regardless of the pattern's type.

**Comment by lahodaj** [comment]
> nit: we could probably add this test case for completeness:
> 
> ```
> class Test {
>     int test() {
>         record R(int a) {}
>         Object o = null;
>         switch (o) {
>             case R(int a) when true: return -1;
>             case R(int a): return 1;
>         }
>     }
> }
> ```

Thanks, done.

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Pull request closed without merging**